modules/reindex/src/main/java/org/opensearch/index/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -403,14 +403,12 @@ void prepareBulkRequest(long thisBatchStartTimeNS, ScrollableHitSource.AsyncResp
      * Send a bulk request, handling retries.
      */
     void sendBulkRequest(BulkRequest request, Runnable onSuccess) {
-        if (logger.isDebugEnabled()) {
-            logger.debug(
-                "[{}]: sending [{}] entry, [{}] bulk request",
-                task.getId(),
-                request.requests().size(),
-                new ByteSizeValue(request.estimatedSizeInBytes())
-            );
-        }
+        logger.debug(
```

Comment:
@akolarkunnu Thanks for raising this PR! From code readability and simplicity point of view your improvement definitely looks better but not having the if block will now lead to creation of new string and runnable objects. Runnable objects can be lightweight depending upon whether they are referencing any variables outside of their scope but string objects may or may not be lightweight and they may or may not be created every time.
I am not in strong disagreement of this change but I would like to see how JVM and CPU graphs look like with some benchmarks before and after this change.

Reply:
Hi @vikasvb90 This change achieves the same lazy logging without having to check for logger levels.
Javadoc of "void debug(String message, Supplier<?>... paramSuppliers)"  says "Logs a message with parameters which are only to be constructed if the logging level is the DEBUG level."
Are you saying to run any specific existing Microbenchmark test case or write a new Microbenchmark test case which covers this changed code area and execute that?

Reply:
The message is very clear that only the log message is constructed later but the arguments which you are now passing will still occupy space in memory and most of the space would be occupied by string objects. So, you need to first check in total what is the impact of this change on heap. 

I am not referring to micro benchmarks but macro benchmarks. https://github.com/opensearch-project/opensearch-benchmark

Reply:
Hi @vikasvb90 

I don't think we will see any quantifiable impact in benchmarking because:
1. Overhead of logging is considerably less than indexing/search activities against where we will benchmark them.
2. In our existing code, we already check for isDebugEnabled/isTraceEnabled at most places, so the code path is already short-circuited. It is the only extra boolean check of isDebugEnabled/isTraceEnabled that we will avoid. 

If we had the debug/trace logs without the boolean isDebugEnabled/isTraceEnabled checks, we may have seen some improvement, but in this case, we will not. This refactoring is intended more from a cleaner code and aiming towards best coding practices, and does not contribute significantly towards performance gains.

Reply:
I am not talking about the overhead of logging itself but the attempt of logging which gets discarded later. I understand what it is intended for. To reiterate, boolean checks earlier prevented the construction of objects passed in log but now they will start occupying space and compute as a result of their construction. Also, it isn't about how big the impact is as compared to search or indexing but whether there is any impact at all and I believe there is. 
If you really want this change to be pushed, at least lets first correctly measure the impact of this change on heap.

Reply:
@akolarkunnu I think some benchmarking results with profiling visualizations will help establish confidence on the changes although this is recommended by log4j: https://logging.apache.org/log4j/2.x/manual/api.html#java-8-lambda-support-for-lazy-logging

We can probably run `geonames` workload on a single node cluster with these configutations:
- no changes, debug & trace disabled
- your changes, debug & trace disabled
AND
- no changes, debug & trace enabled
- your changes, debug & trace enabled


Alternatively, you can reuse the exercise I pointed in https://github.com/opensearch-project/OpenSearch/pull/14723#issuecomment-2289839653 as well - that should also get you CPU profiling. AFAIK, I know ClusterState/Metadata classes have considerate amount of debug/trace logs - verify once.